core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/async/SplittingTransformerTest.java

```diff
@@ -0,0 +1,300 @@
+                future.whenComplete((r, e) -> {
+                    if (e != null) {
+                        fail(e);
+                    }
+                });
+                transformer.onResponse(new TestResultObject("container msg: " + total));
+                transformer.onStream(AsyncRequestBody.fromString(String.format("This is the body of %d.", total)));
+                total++;
+                subscription.request(1);
+            }
+
+            @Override
+            public void onError(Throwable t) {
+                fail("Unexpected onError", t);
+            }
+
+            @Override
+            public void onComplete() {
+                // do nothing, test only
+            }
+        });
+        future.join();
+        String expected = "This is the body of 0.This is the body of 1.This is the body of 2.This is the body of 3.";
+        assertThat(upstreamTestTransformer.contentAsString()).isEqualTo(expected);
+    }
+
+    @Test
+    void whenSubscriberFailsAttempt_UpstreamTransformerCompletesExceptionally() {
+        UpstreamTestTransformer upstreamTestTransformer = new UpstreamTestTransformer();
+        CompletableFuture<Object> future = new CompletableFuture<>();
+        SplittingTransformer<TestResultObject, Object> split =
+            new SplittingTransformer<>(upstreamTestTransformer, 1024*1024*32, future);
+        split.subscribe(new Subscriber<AsyncResponseTransformer<TestResultObject, TestResultObject>>() {
+            private Subscription subscription;
+            private int total = 0;
+
+            @Override
+            public void onSubscribe(Subscription s) {
+                this.subscription = s;
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(AsyncResponseTransformer<TestResultObject, TestResultObject> transformer) {
+                if (total > 2) {
+                    fail("Did not expect more than 2 request to be made");
+                }
+
+                if (total == 2) {
+                    transformer.exceptionOccurred(new RuntimeException("TEST ERROR " + total));
+                    return;
+                }
+
+                transformer.prepare();
+                transformer.onResponse(new TestResultObject("container msg: " + total));
+                transformer.onStream(AsyncRequestBody.fromString(String.format("This is the body of %d.", total)));
+                total++;
+                subscription.request(1);
+            }
+
+            @Override
+            public void onError(Throwable t) {
+                // do nothing, test only
+            }
+
+            @Override
+            public void onComplete() {
+                // do nothing, test only
+            }
+        });
+        assertThatThrownBy(future::join).hasMessageContaining("TEST ERROR 2");
+    }
+
+    @Test
+    void whenDataExceedsBufferSize_UpstreamShouldReceiveAllData() {
```

Comment:
Can you refactor this test to be more descriptive? The title is clear, but the code is filled with numbers that don't explain the role each play - as a reader, taking the time to go through and do the math is a bit much
~~~
int bodyLength = 7*1024; <- why this choice and how does it relate to 16*1024?
int amount = 9; <- why this number
SplittingTransformer<TestResultObject, Object> split =
        new SplittingTransformer<>(upstreamTestTransformer, 16*1024, future); <- break out into it's own variable

//Can you break out this function to it's own method? And call it contentPerTransformer (or a better name)
String content =
   IntStream.range(0, bodyLength).mapToObj(i -> String.valueOf(total)).collect(Collectors.joining());

// Again this could reuse that function and make it clear what the expected result is
for (int i = 0; i < amount; i++) {
        int value = i;
        expected.append(IntStream.range(0, bodyLength).mapToObj(j -> String.valueOf(value)).collect(Collectors.joining()));
    }
        
~~~

Reply:
Added comments and renamed some of the varaible. The idea was to test uneven body size compared to the buffer size